### PR TITLE
goal-conditioned sampler learning

### PR DESCRIPTION
**Pull request by tomsilver**

### CODE REVIEW

File: tests/test_structs.py [hunk 1]
```diff
@@ -745,6 +745,15 @@ def test_segment():
     segment.set_option(option)
     assert segment.has_option()
     assert segment.get_option() == option
+    # Test adding and accessing goals to segments.
```
Comment (ronuchit) [inline tests/test_structs.py:748]: Test adding goals to segments and accessing them.

File: tests/nsrt_learning/test_sampler_learning.py [hunk 2]
```diff
@@ -111,3 +113,44 @@ def test_learn_samplers_failure():
...
+    goal = task.goal
+
+    rng = np.random.default_rng(0)
+    state = task.init
+    goal = task.goal
+    goal_obj = next(iter(goal)).objects[0]
+    objects = [goal_obj]
+
+    input_size = 11
+    classifier = MLPClassifier(input_size, 10000)
```
Comment (ronuchit) [inline tests/nsrt_learning/test_sampler_learning.py:137]: How is the speed of this test? 10000 looks scarily large.

File: src/nsrt_learning/sampler_learning.py [hunk 3]
```diff
@@ -169,7 +188,8 @@ def _create_sampler_data(
     preconditions: Set[LiftedAtom], add_effects: Set[LiftedAtom],
     delete_effects: Set[LiftedAtom], param_option: ParameterizedOption,
     datastore_idx: int
-) -> Tuple[List[Tuple[State, Dict[Variable, Object], _Option]], ...]:
+) -> Tuple[List[Tuple[State, Dict[Variable, Object], _Option,
+                      Optional[Set[GroundAtom]]]], ...]:
```
Comment (ronuchit) [inline src/nsrt_learning/sampler_learning.py:192]: This is getting a little crazy... let's maybe add this to bottom of structs.py?
`SamplerDatapoint = Tuple[State, Dict[Variable, Object], _Option, Optional[Set[GroundAtom]]]`

and then we can change this return type to:
`Tuple[List[SamplerDatapoint], List[SamplerDatapoint]]`
which would further eliminate the unnecessarily vague use of `...` here.

File: src/nsrt_learning/sampler_learning.py [hunk 4]
```diff
@@ -147,11 +158,19 @@ def _learn_neural_sampler(datastores: List[Datastore], nsrt_name: str,
     print("Fitting regressor...")
     X_regressor: List[List[Array]] = []
     Y_regressor = []
-    for state, sub, option in positive_data:  # don't use negative data!
+    for state, sub, option, goal in positive_data:  # don't use negative data!
         # input is state features
         X_regressor.append([np.array(1.0)])  # start with bias term
         for var in variables:
             X_regressor[-1].extend(state[sub[var]])
+        # See comment above.
```
Comment (ronuchit) [inline src/nsrt_learning/sampler_learning.py:166]: There are many comments above. Be more clear, e.g:
"Above, we made the assumption that (blah), which must also be used when assembling data for the regressor."

File: tests/approaches/test_nsrt_learning_approach.py [hunk 5]
```diff
@@ -190,3 +191,13 @@ def test_grammar_search_invention_approach():
                    try_solving=False,
                    sampler_learner="random",
                    additional_settings=additional_settings)
+
+
+def test_sampler_learning_with_goals():
```
Comment (ronuchit) [inline tests/approaches/test_nsrt_learning_approach.py:196]: verify speed of this test (`pytest tests/approaches/test_nsrt_learning_approach.py --durations=0`)

File: tests/nsrt_learning/test_sampler_learning.py (hunk 2 above)
Comment (ronuchit) [reply tests/nsrt_learning/test_sampler_learning.py:137]: `pytest tests/nsrt_learning/test_sampler_learning.py --durations=0`

**Review by ronuchit — COMMENTED** [review]
great to see this goal-conditioned stuff happening! just some small suggestions.

File: tests/nsrt_learning/test_sampler_learning.py (hunk 2 above)
Comment (czeng6) [reply tests/nsrt_learning/test_sampler_learning.py:137]: The classifier isn't used for anything other than to initialize the _LearnedSampler, so the speed of the test isn't really affected (~2.3s). Should I change it to be smaller anyway?

File: tests/approaches/test_nsrt_learning_approach.py (hunk 5 above)
Comment (czeng6) [reply tests/approaches/test_nsrt_learning_approach.py:196]: I get around 10.4s for this test

File: tests/nsrt_learning/test_sampler_learning.py (hunk 2 above)
Comment (ronuchit) [reply tests/nsrt_learning/test_sampler_learning.py:137]: do you mean the speed of just this one function you added, or the entire `test_sampler_learning.py`?

File: tests/approaches/test_nsrt_learning_approach.py (hunk 5 above)
Comment (ronuchit) [reply tests/approaches/test_nsrt_learning_approach.py:196]: do you mean the speed of just this one function you added, or the entire `test_nsrt_learning_approach.py`?

File: tests/test_structs.py (hunk 1 above)
Comment (czeng6) [reply tests/test_structs.py:748]: fixed, thanks!

File: src/nsrt_learning/sampler_learning.py (hunk 3 above)
Comment (czeng6) [reply src/nsrt_learning/sampler_learning.py:192]: done, thanks!

File: src/nsrt_learning/sampler_learning.py (hunk 4 above)
Comment (czeng6) [reply src/nsrt_learning/sampler_learning.py:166]: fixed!

File: tests/approaches/test_nsrt_learning_approach.py (hunk 5 above)
Comment (czeng6) [reply tests/approaches/test_nsrt_learning_approach.py:196]: I meant the entire file; the one function added takes ~2.4s

File: tests/nsrt_learning/test_sampler_learning.py (hunk 2 above)
Comment (czeng6) [reply tests/nsrt_learning/test_sampler_learning.py:137]: Hmm sorry, 2.3s is the speed of the entire file. Seems like without the new function the file takes ~2.1s to run.

File: tests/approaches/test_nsrt_learning_approach.py (hunk 5 above)
Comment (ronuchit) [reply tests/approaches/test_nsrt_learning_approach.py:196]: okay, not a problem then!

File: tests/nsrt_learning/test_sampler_learning.py (hunk 2 above)
Comment (ronuchit) [reply tests/nsrt_learning/test_sampler_learning.py:137]: okay not a problem

**Review by ronuchit — APPROVED** [review]
LGTM, but let @tomsilver check as well! Great stuff Catherine!

**Review by tomsilver — COMMENTED** [review]
Approved! (I can't do it officially since I opened the PR.) Merge when ready @czeng6 , thank you!

**Pull request merged**